src/mono/wasm/runtime/dotnet-crypto-worker.js

```diff
@@ -0,0 +1,170 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+var ChannelWorker = {
+    _impl: class {
+        // BEGIN ChannelOwner contract - shared constants.
+        get STATE_IDX() { return 0; }
+        get MSG_SIZE_IDX() { return 1; }
+
+        // Communication states.
+        get STATE_SHUTDOWN() { return -1; } // Shutdown
+        get STATE_IDLE() { return 0; }
+        get STATE_REQ() { return 1; }
+        get STATE_RESP() { return 2; }
+        get STATE_REQ_P() { return 3; } // Request has multiple parts
+        get STATE_RESP_P() { return 4; } // Response has multiple parts
+        get STATE_AWAIT() { return 5; } // Awaiting the next part
+        // END ChannelOwner contract - shared constants.
+
+        constructor(comm_buf, msg_buf, msg_char_len) {
+            this.comm = new Int32Array(comm_buf);
+            this.msg = new Uint16Array(msg_buf);
+            this.msg_char_len = msg_char_len;
+        }
+
+        async await_request(async_call) {
+            for (;;) {
+                // Wait for signal to perform operation
+                Atomics.wait(this.comm, this.STATE_IDX, this.STATE_IDLE);
+
+                // Read in request
+                var req = this._read_request();
+                if (req === this.STATE_SHUTDOWN)
+                    break;
+
+                var resp = null;
+                try {
+                    // Perform async action based on request
+                    resp = await async_call(req);
+                }
+                catch (err) {
+                    console.log("Request error: " + err);
+                    resp = JSON.stringify(err);
```

Comment:
Should there be a way to indicate that this is returning an error, or that the operation failed? Do IIUC, that the caller will get the response, and if the length happens to match the expected one, then it will think that the error response is actually valid?